File: Yr2022/day_seventeen.py
from time import perf_counter_ns
from tqdm import tqdm
from dataclasses import dataclass
from typing import List, Dict, Tuple, DefaultDict, FrozenSet
import functools
import itertools

import numpy as np

import matplotlib.pyplot as plt



CACHE_DEPTH=10

@dataclass
class TetrisShape:
    #origin is bottom left
    origin: Tuple[int, int]
    width: int
    height: int
    arr: np.array

    def set_depth(self, start_depth):
        self.origin = (start_depth, 2)

    def wind_shift(self, board_arr, right):
        result_str = "WIND RIGHT " if right else "WIND LEFT"
        if right and self.origin[1]+self.width <= board_arr.shape[1]-1:
            if not self.collides(board_arr, xshift=+1, yshift=-1):
                #self.origin[0] += 1
                self.origin = (self.origin[0], self.origin[1] + 1)
                result_str += " succesful"
            else:
                result_str += " collided"
        elif right:
                result_str += " bounced"
        if not right and self.origin[1] > 0:
            if not self.collides(board_arr, xshift=-1, yshift=-1):
                #self.origin[0] -= 1
                self.origin = (self.origin[0], self.origin[1] - 1)
                result_str += " successful"
            else:
                result_str += " collided"
        elif not right:
                result_str += " bounced"
        return result_str

    def fall(self):
        self.origin = (self.origin[0] + 1, self.origin[1])

    def board_sub_arr_idx(self, board_arr):
        comp_arr_width = min(self.arr.shape[1], board_arr.shape[1] - self.origin[1])
        ymin, ymax = self.origin[0]-2, self.origin[0]+2
        #ymin, ymax = self.origin[0]-3+yshift, self.origin[0]+1+yshift
        #xmin, xmax = self.origin[0], self.origin[0]+comp_arr_width
        xmin, xmax = self.origin[1], self.origin[1]+self.width
        return ymin, ymax, xmin, xmax

    def shape_x_idx(self, board_arr):
        #shape_xmax = min(self.arr.shape[0], board_arr.shape[0]-self.origin[0])
        shape_xmax = self.origin[1] + self.width
        return shape_xmax

    def collides(self, board_arr, xshift=0, yshift=0):
        #board_width = board_arr.shape[0]
        #comp_arr_width = min(self.arr.shape[0], board_width - self.origin[0])
        #sub_arr = board_arr[self.origin[1]-3:self.origin[1]+2][self.origin[0]:self.origin[0]+4]
        #ymin, ymax = self.origin[1]-2, self.origin[1]+2
        #xmin, xmax = self.origin[0], self.origin[0]+comp_arr_width
        ymin, ymax, xmin, xmax = self.board_sub_arr_idx(board_arr)
        ymin, ymax = ymin+yshift, ymax+yshift
        xmin, xmax = xmin+xshift, xmax+xshift
        #shape_xmax = self.shape_x_idx(board_arr)
        board_sub_arr = board_arr[ymin:ymax,xmin:xmax]
        
        #shape_sub_arr = self.arr[:, 0:shape_xmax]
        shape_sub_arr = self.arr[:, 0:self.width]

        #print("ORIGIN", self.origin, ymin, ymax, xmin, xmax, shape_xmax)
        #print("BOARD ARR", board_sub_arr, "->", board_sub_arr.shape)
        #print("SHAPE ARR", shape_sub_arr, "->", shape_sub_arr.shape)
        #print("-------------------------")
        """
        #sub_arr = board_arr[8087][0]
        
        print("BOARD ARR", board_arr[self.origin[1]][:])
        
        
        print("-------------------------")
        print("-------------------------")
        print("-------------------------")
        print("SHAPE ARR ", self.arr)
        print("-------------------------")
        print("-------------------------")
        print("-------------------------")
        print("-------------------------")
        """

        collide = np.logical_and(shape_sub_arr, board_sub_arr).sum() > 0
        #if collide:
        #    print(shape_sub_arr)
        #    print("------------------------")
        #    print(board_sub_arr)
        return collide


class Long(TetrisShape):
    def __init__(self):
        self.width, self.height = 4, 1
        self.arr = np.array([[0,0,0,0],[0,0,0,0],[0,0,0,0],[1,1,1,1]])

class Tall(TetrisShape):
    def __init__(self):
        self.width, self.height = 1, 4
        self.arr = np.array([[1,0,0,0],[1,0,0,0],[1,0,0,0],[1,0,0,0]])

class Cross(TetrisShape):
    def __init__(self):
        self.width, self.height = 3, 3
        self.arr = np.array([[0,0,0,0],[0,1,0,0],[1,1,1,0],[0,1,0,0]])

class Corner(TetrisShape):
    def __init__(self):
        self.width, self.height = 3, 3
        self.arr = np.array([[0,0,0,0],[0,0,1,0],[0,0,1,0],[1,1,1,0]])

class Square(TetrisShape):
    def __init__(self):
        self.width, self.height = 2, 2
        self.arr = np.array([[0,0,0,0],[0,0,0,0],[1,1,0,0],[1,1,0,0]])




class TetrisBoard:
    def __init__(self, winds:List[bool], board_width:int = 7, board_depth: int = 8088):
        self.arr =  np.zeros(shape=(board_depth, board_width))
        #with np.nditer(self.arr, op_flags=['readwrite']) as it:
        #    for idx, x in enumerate(it):
        #        x[...] = idx

        #set floor
        self.arr[board_depth-1] = np.ones(shape=(board_width))
                
        self.top_line: int = board_depth-2
        #self.winds = itertools.cycle(winds)
        self.winds = winds
        self.wind_idx = 0
        #self.shapes = itertools.cycle([Long(), Cross(), Corner(), Tall(), Square()])
        self.shapes = [Long(), Cross(), Corner(), Tall(), Square()]
        self.shape_idx = 0

        #print("Initialised", self.shapes)
        
    @property
    def hash_wind(self):
        overlap = CACHE_DEPTH - (len(self.winds) - 1 - self.wind_idx)
        if overlap >= 0:
            result = self.winds[self.wind_idx:].extend(self.winds[0:overlap])
        else:
            result = self.winds[self.wind_idx: self.wind_idx+CACHE_DEPTH]
        result_hash = 0
        for ele_idx, ele in enumerate(result):
            val = 1 if ele else 0
            result_hash = result_hash + (val << ele_idx)
        self.wind_idx = self.wind_idx + 1 if self.wind_idx < len(self.winds)-1 else 0
        return result_hash

    @property
    def wind(self):
        result = self.winds[self.wind_idx]
        self.wind_idx = self.wind_idx + 1 if self.wind_idx < len(self.winds)-1 else 0
        return result

    @property
    def shape(self):
        result = self.shapes[self.shape_idx]
        self.shape_idx = self.shape_idx + 1 if self.shape_idx < len(self.shapes)-1 else 0
        return result

    def shape_stop(self, shape):
        ymin, ymax, xmin, xmax = shape.board_sub_arr_idx(self.arr)
        #shape_xmin, shape_xmax = shape.origin[0] , self.arr.shape[0]-shape.origin
        shape_xmax = shape.shape_x_idx(self.arr)
        self.arr[ymin-1:ymax-1,xmin:xmax] = np.logical_or(self.arr[ymin-1:ymax-1,xmin:xmax],
                                                         shape.arr[:, 0:shape.width ])

        #self.arr[shape.origin[1]-4:shape.origin[1],
        #        shape.origin[0]:shape.origin[0]+shape.width+1] = shape.arr
        old_top_line = self.top_line
        self.top_line = min(self.top_line, shape.origin[0] - shape.height)
        return old_top_line - self.top_line



    def shape_stop_fast(self, shape):
        ymin, ymax, xmin, xmax = shape.board_sub_arr_idx(self.arr)
        
        #shape_xmax = shape.shape_x_idx(self.arr)
        self.arr[ymin-1:ymax-1,xmin:xmax] = np.logical_or(self.arr[ymin-1:ymax-1,xmin:xmax],
                                                         shape.arr[:, 0:shape.width ])

        new_bottom = self.arr.shape[0]
        for x_idx in range(0, self.arr.shape[1]):
            new_bottom = min(new_bottom, (self.arr[:, x_idx] != 0).argmax(axis=0))
        shift = self.arr.shape[0] - new_bottom
        self.arr = np.roll(self.arr, shift, axis=0)
        #old_top_line = self.top_line
        #self.top_line = min(self.top_line, shape.origin[0] - shape.height)
        self.top_line = min(self.top_line, shape.origin[0] - shape.height) + shift
        #return old_top_line - self.top_line
        return shift

    def add_shape_plot(self, shape_number=0):
        
        #shape = next(self.shapes)
        shape = self.shape[self.shape_idx]
        shape.set_depth(self.top_line - 3)
        not_collided, turn_idx = True, 0
        
        while (not_collided):
            #self.plot(shape, shape_idx, turn_idx, 'MOVE')
            turn_idx += 1
            r_str = shape.wind_shift(right = self.wind, board_arr=self.arr)
            #self.plot(shape, shape_idx, turn_idx, r_str)
            turn_idx += 1            
            if shape.collides(self.arr):
                not_collided = False
            else:
                shape.fall()
        
        shift = self.shape_stop(shape)
        self.plot(shape, shape_number, turn_idx, 'STOP ' + str(shift) )
        return shift

    def add_shape(self, shape): # shape_number=0):
        #hash_arr = tuple(tuple(row) for row in self.arr.astype(int))
        #hash_wind = self.hash_wind(self.wind_idx)
        #print(hash_arr)
        
        #shape = next(self.shapes)
        #shape = self.shape[self.shape_idx]
        shape.set_depth(self.top_line - 3)
        not_collided, turn_idx = True, 0
        
        while (not_collided):
            
            r_str = shape.wind_shift(right = self.wind, board_arr=self.arr)
            
            if shape.collides(self.arr):
                not_collided = False
            else:
                shape.fall()
        
        shift = self.shape_stop(shape)
        
        return tuple(tuple(row) for row in self.arr.astype(int)), shift





    def shift_arr(self, shift=50):
        #self.arr[self.arr.shape[0]//2:] = self.arr[:self.arr.shape[0]//2]
        #self.arr[:self.arr.shape[0]//2] = np.zeros(shape=(self.arr.shape[0]//2, self.arr.shape[1]))
        #return self.arr.shape[0]//2
        self.arr = np.roll(self.arr, shift, axis=0)
        self.top_line += shift

    def play(self, num_shapes: int = 2022, plot=False):
        
        
        shifted = []
        for i in tqdm(range(0, num_shapes)):
            plot  = True if 990<i<1005 else False
            #shift = self.add_shape_plot(i, plot)
            shift = self.add_shape(i, plot)
            if shift > 0:
                #print(shift)
                #shifted += shift
                shifted.append(shift)
                if i > 10:
                    self.shift_arr(shift)
            
        print(self.arr.shape[0], self.top_line)
        print(shifted)
        print(sum(shifted))
        return self.arr.shape[0] - self.top_line -2


        
    def play_fast(self, num_shapes: int = 2022):
        
        state_cache = {}
        state = tuple(tuple(row) for row in self.arr.astype(int))
        shifter = 0
        
        for i in tqdm(range(0, num_shapes)):
            
            wind_hash = self.hash_wind
            shape = self.shape
            shape_idx = self.shape_idx
            
            if (shape_idx, wind_hash, state) in state_cache:
                state, shift = state_cache[(shape_idx, wind_hash, state)]

            else:
                state, shift = self.add_shape(shape)
                state_cache[(shape_idx, wind_hash, state)] = state, shift
            
            shifter += shift
            
        print("SHIFTER ", shifter)
        return shifter

    def plot(self, shape, shape_idx, turn_idx, title='BLANK'):
        fig,ax = plt.subplots(figsize=(10,8))
        #print(self.arr.sum(), self.)
        #print(np.sum(self.arr < 0))
        plot_arr = np.copy(self.arr)
        ymin, ymax, xmin, xmax = shape.board_sub_arr_idx(plot_arr)
        #shape_xmin, shape_xmax = shape.origin[0] , self.arr.shape[0]-shape.origin
        #shape_xmax = shape.shape_x_idx(plot_arr)
        plot_arr[ymin-1:ymax-1,xmin:xmax] = plot_arr[ymin-1:ymax-1,xmin:xmax] + \
                                                    (-1) * shape.arr[:, 0:shape.width ]
        ax.imshow(plot_arr[-100:,:])
        ax.set_title(title)
        dir = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\vis\tetris"
        out_img_name = f"{dir}_{shape_idx}_{turn_idx}.png"
        fig.savefig(out_img_name)

        






def play_tetris(filename: str):
    cmd_lines= []
    with open(filename, 'r') as file:
        cmd_lines = file.readlines()

    example_line = ">>><<><>><<<>><>>><<<>>><<<><<<>><>><<>>"        

    #wind_list = [char == '>' for char in cmd_lines[0].strip()]
    wind_list = [char == '>' for char in example_line]
    #for char in cmd_lines[0].strip():

    debug = True
    do_plots = False
    #num_shapes = 2022
    num_shapes = 1_000_000_000_000

    if debug:
        num_shapes = 5
        do_plots = True


    board = TetrisBoard(winds=wind_list, board_depth=10)
        
    #part_one = board.play(num_shapes=num_shapes, plot=do_plots)
    part_one = board.play_fast(num_shapes=num_shapes)
    
    return part_one, None


  

if __name__ == "__main__":
    time_start = perf_counter_ns()
    file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\puzzle_day_seventeen.txt"     
    
    part_one, part_two = play_tetris(file_loc)
    
    print(f"(Part One): {part_one}     (Part Two): {part_two}")   
    print(f"Elapsed time : {(perf_counter_ns() - time_start)/1000:.0f} microseconds")  

File: Yr2022/src/day_seventeen.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;
use std::collections::HashMap;


fn main() {   
    let start = Instant::now();
    let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_seventeen.txt";
    let (part_one, part_two) = solve_tetris(file_name);
    let duration = start.elapsed();    
    println!("Part One: {}  Part Two: {}", part_one, part_two);
    println!("Time taken to execute: {} microseconds", duration.as_micros());    
}



const NUM_SHAPES: usize = 5;
const SHAPE_HEIGHT: usize = 4;
const SHAPE_WIDTH: [usize; NUM_SHAPES] = [4,3,3,1,2];
const SHAPE_DATA: [[u8; SHAPE_HEIGHT]; NUM_SHAPES] = [
    [0b1111,0b0000,0b0000,0b0000], //Tall
    [0b0010, 0b0111, 0b0010, 0b0000], // Cross
	[0b0111, 0b0100, 0b0100, 0b0000],	// 
	[0b0001, 0b0001, 0b0001, 0b0001],	// Long
	[0b0011, 0b0011, 0b0000, 0b0000], // Square
];
const BOARD_WIDTH: usize = 7;
const BOARD_DEPTH: usize = 128;

struct TetrisBoard{ 
    height_from_floor:usize,
    height_from_cutoff: usize,
    height_of_cutoff: usize,
    board: [u8; BOARD_DEPTH],
}

impl Default for TetrisBoard {
    fn default() -> TetrisBoard {
        TetrisBoard {height_from_floor:0, height_from_cutoff: 0, height_of_cutoff: 0, board: [0; BOARD_DEPTH],}
    }
}

impl TetrisBoard {
    pub fn no_collision(&self, shape: [u8; SHAPE_HEIGHT], x: usize, y:usize) -> bool {
        (0..SHAPE_HEIGHT).all(|idx| ((shape[idx] << x) & self.board[y+idx]) == 0)
    }
    pub fn place_shape(&mut self, shape: [u8; SHAPE_HEIGHT], x: usize, y:usize) {
        for _i in 0..SHAPE_HEIGHT {
            let vert_slice = shape[_i] << x;
            if vert_slice != 0 {
                self.board[y+_i] |= vert_slice;
                let height = y + _i +1;
                if height > self.height_from_cutoff {
                    self.height_from_cutoff = height;
                    self.height_from_floor = self.height_of_cutoff + height;
                }
            }
        }
    }
    pub fn find_cutoff(&self) -> usize {
        let mask = ((1<< BOARD_WIDTH) -1) as u8;
        let mut row_mask = mask;
        for y in (1..self.height_from_cutoff).rev() {
            row_mask = row_mask & !self.board[y];
            for _i in 0..BOARD_WIDTH {
                let left_wall_mask = (row_mask >> 1) & mask & !self.board[y];
                let right_wall_mask = (row_mask << 1) & mask & !self.board[y];
                let collision_mask = row_mask | left_wall_mask | right_wall_mask;
                if collision_mask == row_mask { break; } // No collision
                row_mask = collision_mask;
            }
            if row_mask == 0 { return y;}    
        }
        
        return 0;
    }
    pub fn make_cutoff(&mut self){
        let y_cutoff = self.find_cutoff();
        self.board.copy_within(y_cutoff.., 0); // shift by y
        self.height_from_cutoff -= y_cutoff;
        self.height_of_cutoff += y_cutoff;
        self.board[self.height_from_cutoff..].fill(0); //fill upper rows with zero
    }
}


#[derive(Clone, PartialEq, Eq, Hash)]
struct State {
    wind_idx: usize,
    board: [u8; BOARD_DEPTH],
}

#[derive(Clone, Copy)]
struct Result {
    num_shapes_placed: usize,
    total_height: usize,
}

fn play_tetris(tetris: &mut TetrisBoard, winds: &[u8], mut wind_idx: usize, num_shapes: usize) {

    let wind_len = winds.len();
    let mut cache: HashMap<State, Result> = HashMap::new();
    let mut shape_idx = 0;

    for _i in 0..num_shapes {
        let shape = SHAPE_DATA[shape_idx];
        let shape_width = SHAPE_WIDTH[shape_idx];
        let mut x_pos = 2;
        let mut y_pos = tetris.height_from_cutoff + 3;

        if y_pos + SHAPE_HEIGHT >= BOARD_DEPTH { // if overflow the top, shift down
            tetris.make_cutoff();
            y_pos = tetris.height_from_cutoff + 3;
        }

        loop {
            let wind = winds[wind_idx];
            wind_idx = (wind_idx + 1)% wind_len;

            match wind {
                b'>' if x_pos + shape_width >= BOARD_WIDTH => (),  //hit right wall
                b'>' => { if tetris.no_collision(shape, x_pos+1, y_pos){x_pos+=1;}}, //wind shift right
                b'<' if x_pos == 0 => (),  //hit left wall
                b'<' => {if tetris.no_collision(shape, x_pos-1, y_pos){x_pos-=1;}} // wind shift left
                _ => unreachable!(),
            }
            if y_pos == 0 {break;} // hit floor
            if !tetris.no_collision(shape, x_pos, y_pos-1) { break; } // hit other shape
            y_pos -= 1;
        }

        tetris.place_shape(shape, x_pos, y_pos);
        shape_idx = (shape_idx + 1 ) % NUM_SHAPES;


        let num_shapes_placed =_i+1;
        let cache_depth = 10 * wind_len * NUM_SHAPES;
        if num_shapes > cache_depth && (num_shapes_placed % cache_depth) == 0 {
            let current_state = State { wind_idx, board: tetris.board.clone(),};
            let current_result = Result { num_shapes_placed, total_height:tetris.height_from_floor,};

            if let Some(previous_result) = cache.get(&current_state) {
                let extra_shapes = current_result.num_shapes_placed - previous_result.num_shapes_placed;
                let extra_height = current_result.total_height - previous_result.total_height;
                let mut shapes_remaining = num_shapes - current_result.num_shapes_placed;
                let num_skips = shapes_remaining / extra_shapes;
                shapes_remaining = shapes_remaining % extra_shapes;
                
                tetris.height_from_floor += num_skips*extra_height;
                tetris.height_of_cutoff += num_skips*extra_height;
                return play_tetris(tetris, winds, wind_idx, shapes_remaining);
            }
            else {
                cache.insert(current_state, current_result);
            }
        }

        

    }

}








fn solve_tetris(my_file_name: &str) -> (usize, usize) {
    let file = File::open(my_file_name).expect("File does not exist");
    let buf_lines = BufReader::new(file).lines();
    let line_vec: Vec<String> = buf_lines.map(|l| l.expect("Fail to parse")).collect();

    let winds: &str = line_vec[0].trim();


    let mut board = TetrisBoard::default();
    play_tetris(&mut board, winds.as_bytes(), 0, 2022);
    let part_one = board.height_from_floor;

    let mut board_two = TetrisBoard::default();
    play_tetris(&mut board_two, winds.as_bytes(), 0, 1000000000000);
    let part_two = board_two.height_from_floor;


    return (part_one, part_two);
}
